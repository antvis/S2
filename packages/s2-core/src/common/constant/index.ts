export * from './events';
export * from './basic';
export * from './classnames';
export * from './condition';
export * from './dataConfig';
export * from './frozen';
export * from './interaction';
export * from './options';
export * from './scroll';
export * from './theme';
export * from './tooltip';
export * from './resize';
