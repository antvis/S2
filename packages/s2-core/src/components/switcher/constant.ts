import { ColIcon, RowIcon, ValueIcon } from '../icons/index';

export enum FieldType {
  Rows = 'rows',
  Cols = 'cols',
  Values = 'values',
}

export const SWITCHER_CONFIG = {
  [FieldType.Rows]: {
    text: '行头',
    icon: RowIcon,
  },
  [FieldType.Cols]: {
    text: '列头',
    icon: ColIcon,
  },
  [FieldType.Values]: {
    text: '值',
    icon: ValueIcon,
  },
} as const;
