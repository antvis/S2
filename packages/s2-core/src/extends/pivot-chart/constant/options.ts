import type { S2Options } from '@antv/s2';

export const FIXED_OPTIONS: S2Options = {
  hierarchyType: 'grid',
};
