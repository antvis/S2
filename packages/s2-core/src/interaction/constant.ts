export enum InteractionNames {
  BRUSH_SELECTION_INTERACTION = 'spreadsheet:brush-selection',
  COL_ROW_RESIZE_INTERACTION = 'spreadsheet:row-col-resize',
  DATACELL_MUTI_SELECTION_INTERACTION = 'spreadsheet:datacell-muti-selection',
  COL_ROW_MUTI_SELECTION_INTERACTION = 'spreadsheet:col-row-muti-selection',
}

export enum EventNames {
  DATACELL_CLICK_EVENT = 'spreadsheet:data-cell-click',
  CORNER_TEXT_CLICK_EVENT = 'spreadsheet:corner-text-click',
  ROW_COLUMN_CLICK_EVENT = 'spreadsheet:row-column-click',
  ROW_TEXT_CLICK_EVENT = 'spreadsheet:row-text-click',
  HOVER_EVENT = 'spreadsheet:hover',
}