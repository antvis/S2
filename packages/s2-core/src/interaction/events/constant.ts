export const KEEP_HOVER_TIME = 800;
