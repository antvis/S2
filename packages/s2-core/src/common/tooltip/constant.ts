import { PRE_CLASS } from '@/common/constant/classnames';

export const TOOLTIP_CLASS_PRE = `${PRE_CLASS}-tooltip`;

export const TOOLTIP_OPERATION_CLASS_PRE = `${TOOLTIP_CLASS_PRE}-operation`;

export const DEFAULT_ICON_PROPS = {
  width: 14,
  height: 14,
  style: {
    verticalAlign: 'sub',
    marginRight: 4,
  },
};

export const POSITION_X_OFFSET = 10;
export const POSITION_Y_OFFSET = 10;
