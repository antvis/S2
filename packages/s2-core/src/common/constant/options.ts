import { S2Options } from '../interface/s2Options';
import { ScrollbarPositionType } from '../constant/interaction';
import { EMPTY_PLACEHOLDER } from './basic';
import { Style } from '@/common/interface/basic';
import { ResizeType } from '@/common/constant/resize';

export const MIN_DEVICE_PIXEL_RATIO = 1;

export enum LayoutWidthTypes {
  Adaptive = 'adaptive',
  ColAdaptive = 'colAdaptive',
  Compact = 'compact',
}

export const DEFAULT_STYLE: Readonly<Style> = {
  layoutWidthType: LayoutWidthTypes.Adaptive,
  treeRowsWidth: 120,
  collapsedRows: {},
  collapsedCols: {},
  cellCfg: {
    width: 96,
    height: 30,
  },
  rowCfg: {
    width: null,
    widthByField: {},
    heightByField: {},
  },
  colCfg: {
    height: 30,
    widthByFieldValue: {},
    heightByField: {},
  },
  device: 'pc',
};

export const DEFAULT_OPTIONS: Readonly<S2Options> = {
  width: 600,
  height: 480,
  debug: false,
  hierarchyType: 'grid',
  conditions: {},
  totals: {},
  tooltip: {
    showTooltip: false,
    autoAdjustBoundary: 'body',
    operation: {
      hiddenColumns: false,
      trend: false,
      sort: false,
      menus: [],
    },
  },
  interaction: {
    linkFields: [],
    hiddenColumnFields: [],
    selectedCellsSpotlight: false,
    hoverHighlight: true,
    hoverFocus: true,
    scrollSpeedRatio: {
      horizontal: 1,
      vertical: 1,
    },
    autoResetSheetStyle: true,
    brushSelection: true,
    multiSelection: true,
    rangeSelection: true,
    scrollbarPosition: ScrollbarPositionType.CONTENT,
    resize: {
      rowCellVertical: true,
      cornerCellHorizontal: true,
      colCellHorizontal: true,
      colCellVertical: true,
      rowResizeType: ResizeType.ALL,
    },
  },
  showSeriesNumber: false,
  customSVGIcons: [],
  showDefaultHeaderActionIcon: false,
  headerActionIcons: [],
  style: DEFAULT_STYLE,
  frozenRowHeader: true,
  frozenRowCount: 0,
  frozenColCount: 0,
  frozenTrailingRowCount: 0,
  frozenTrailingColCount: 0,
  hdAdapter: true,
  cornerText: '',
  placeholder: EMPTY_PLACEHOLDER,
  supportCSSTransform: false,
  devicePixelRatio: window.devicePixelRatio,
};
