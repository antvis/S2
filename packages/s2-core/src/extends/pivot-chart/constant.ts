import type { G2Spec } from '@antv/g2';
import { type S2DataConfig, type S2Options } from '@antv/s2';
import { ChartDataCell } from './cell/chart-data-cell';
import { AxisRowColumnClick } from './interaction/axis-click';
import { AxisHover } from './interaction/axis-hover';

export const DEFAULT_G2_SPEC: G2Spec = {
  type: 'interval',
  autoFit: true,
  animate: false,
  axis: false,
  legend: false,
};

export const DEFAULT_G2_AXIS_SPEC: G2Spec = {
  autoFit: true,
  animate: false,
  line: false,
  tick: false,
  grid: false,
};

export const FIXED_OPTIONS: S2Options = {
  hierarchyType: 'grid',
  interaction: {
    selectedCellsSpotlight: false,
    copy: {
      enable: false,
    },
  },
  style: {
    colCell: {
      hideValue: false,
    },
  },
};

export const DEFAULT_OPTIONS: S2Options = {
  chartCoordinate: 'cartesian',
  chartSpec: DEFAULT_G2_SPEC,
  dataCell: (viewMeta, spreadsheet) => new ChartDataCell(viewMeta, spreadsheet),
  interaction: {
    customInteractions: [
      {
        key: 'axisHover',
        interaction: AxisHover,
      },
      {
        key: 'axisClick',
        interaction: AxisRowColumnClick,
      },
    ],
  },
};

export const FIXED_DATA_CONFIG: Partial<S2DataConfig> = {
  fields: {
    customValueOrder: null,
  },
};

export const DEFAULT_MEASURE_SIZE = 200;
export const DEFAULT_ROW_AXIS_SIZE = 100;
export const DEFAULT_COL_AXIS_SIZE = 50;
export const DEFAULT_DIMENSION_SIZE = 50;

/**
 * row axis
 */
export const KEY_GROUP_ROW_AXIS_SCROLL = 'rowAxisScrollGroup';
export const KEY_GROUP_ROW_AXIS_FROZEN = 'rowAxisHeaderFrozenGroup';
export const KEY_GROUP_ROW_AXIS_HEADER_FROZEN_TRAILING =
  'rowAxisHeaderFrozenTrailingGroup';
export const KEY_GROUP_ROW_AXIS_RESIZE_AREA = 'rowAxisHeaderResizeArea';

/**
 * column axis
 */
export const KEY_GROUP_COL_AXIS_SCROLL = 'colAxisScrollGroup';
export const KEY_GROUP_COL_AXIS_FROZEN = 'colAxisFrozenGroup';
export const KEY_GROUP_COL_AXIS_FROZEN_TRAILING = 'colAxisFrozenTrailingGroup';
export const KEY_GROUP_COL_AXIS_RESIZE_AREA = 'colAxisHeaderResizeArea';

export const PLACEHOLDER_FIELD = '$$placeholder$$';
