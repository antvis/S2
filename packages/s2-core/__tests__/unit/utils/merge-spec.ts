import { DEFAULT_STYLE } from '@/common/constant/options';
import {
  customMerge,
  getSafetyDataConfig,
  getSafetyOptions,
} from '@/utils/merge';
import type { S2DataConfig } from '@/common';

describe('merge test', () => {
  test('should replace old array with new one', () => {
    expect(customMerge({ arr: [1, 2, 3] }, { arr: [4, 5, 6] })).toEqual({
      arr: [4, 5, 6],
    });
  });

  test('should force replace empty array', () => {
    expect(customMerge({ arr: [1, 2, 3] }, { arr: [] })).toEqual({
      arr: [],
    });
  });

  test('should return new object', () => {
    const obj = { name: 'name' };
    const result = customMerge(obj, { age: 100 });

    expect(obj).toEqual({ name: 'name' });
    expect(result).toEqual({ name: 'name', age: 100 });
  });

  test('should get safety data config', () => {
    expect(getSafetyDataConfig(null)).toStrictEqual({
      data: [],
      totalData: [],
      fields: {
        rows: [],
        columns: [],
        values: [],
        customTreeItems: [],
        valueInCols: false,
      },
      meta: [],
      sortParams: [],
      filterParams: [],
    });
  });

  test('should cancel valueInCols if customTreeItems is not empty by get safety data config', () => {
    const fields: Partial<S2DataConfig['fields']> = {
      customTreeItems: [{ key: '1', title: 'test' }],
      valueInCols: true,
    };
    expect(
      getSafetyDataConfig({
        fields,
      }),
    ).toStrictEqual({
      data: [],
      totalData: [],
      fields: {
        ...fields,
        rows: [],
        columns: [],
        values: [],
        valueInCols: false,
      },
      meta: [],
      sortParams: [],
      filterParams: [],
    });
  });

  test('should cancel valueInCols if value is empty by get safety data config', () => {
    const fields: Partial<S2DataConfig['fields']> = {
      values: [],
      valueInCols: true,
    };
    expect(
      getSafetyDataConfig({
        fields,
      }),
    ).toStrictEqual({
      data: [],
      totalData: [],
      fields: {
        ...fields,
        rows: [],
        columns: [],
        customTreeItems: [],
        valueInCols: false,
      },
      meta: [],
      sortParams: [],
      filterParams: [],
    });
  });

  test('should get safety options', () => {
    // 加这个测试可以防止 本地跑demo 修改了默认配置 直接提交
    expect(getSafetyOptions(null)).toStrictEqual({
      width: 600,
      height: 480,
      debug: false,
      hierarchyType: 'grid',
      conditions: {},
      cornerText: '',
      totals: {},
      tooltip: {
        showTooltip: false,
        autoAdjustBoundary: 'body',
        operation: {
          hiddenColumns: false,
          trend: false,
          sort: false,
          menus: [],
        },
      },
      interaction: {
        linkFields: [],
        hiddenColumnFields: [],
        selectedCellsSpotlight: false,
        hoverHighlight: true,
        hoverFocus: true,
        scrollSpeedRatio: {
          horizontal: 1,
          vertical: 1,
        },
        autoResetSheetStyle: true,
        brushSelection: true,
        multiSelection: true,
        rangeSelection: true,
        resize: {
          colCellHorizontal: true,
          colCellVertical: true,
          cornerCellHorizontal: true,
          rowCellVertical: true,
          rowResizeType: 'all',
        },
        scrollbarPosition: 'content',
      },
      frozenRowHeader: true,
      showSeriesNumber: false,
      customSVGIcons: [],
      showDefaultHeaderActionIcon: false,
      headerActionIcons: [],
      style: DEFAULT_STYLE,
      frozenRowCount: 0,
      frozenColCount: 0,
      frozenTrailingRowCount: 0,
      frozenTrailingColCount: 0,
      hdAdapter: true,
      placeholder: '-',
      supportCSSTransform: false,
      devicePixelRatio: window.devicePixelRatio,
    });
  });

  test('should get custom options', () => {
    const options = getSafetyOptions({
      tooltip: {
        showTooltip: false,
        operation: {
          sort: false,
          menus: [
            {
              key: 'custom',
              text: 'custom',
            },
          ],
        },
      },
    });
    expect(options.tooltip).toStrictEqual({
      autoAdjustBoundary: 'body',
      operation: {
        hiddenColumns: false,
        trend: false,
        sort: false,
        menus: [
          {
            key: 'custom',
            text: 'custom',
          },
        ],
      },
      showTooltip: false,
    });
  });

  test('should get custom data config', () => {
    const dataConfig = getSafetyDataConfig({
      fields: {
        rows: ['test'],
        values: ['value'],
      },
    });
    expect(dataConfig.fields).toStrictEqual({
      rows: ['test'],
      columns: [],
      values: ['value'],
      customTreeItems: [],
      valueInCols: true,
    });
  });
});
