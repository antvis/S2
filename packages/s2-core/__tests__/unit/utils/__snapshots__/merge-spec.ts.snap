// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge test should get safety options 1`] = `
Object {
  "conditions": Object {},
  "cornerExtraFieldText": "",
  "cornerText": "",
  "customSVGIcons": Array [],
  "debug": false,
  "devicePixelRatio": 2,
  "frozenColCount": 0,
  "frozenRowCount": 0,
  "frozenRowHeader": true,
  "frozenTrailingColCount": 0,
  "frozenTrailingRowCount": 0,
  "hdAdapter": true,
  "headerActionIcons": Array [],
  "height": 480,
  "hierarchyType": "grid",
  "interaction": Object {
    "autoResetSheetStyle": true,
    "brushSelection": Object {
      "col": false,
      "data": true,
      "row": false,
    },
    "eventListenerOptions": false,
    "hiddenColumnFields": Array [],
    "hoverFocus": Object {
      "duration": 800,
    },
    "hoverHighlight": true,
    "linkFields": Array [],
    "multiSelection": true,
    "overscrollBehavior": "auto",
    "rangeSelection": true,
    "resize": Object {
      "colCellHorizontal": true,
      "colCellVertical": true,
      "cornerCellHorizontal": true,
      "rowCellVertical": true,
      "rowResizeType": "all",
    },
    "scrollSpeedRatio": Object {
      "horizontal": 1,
      "vertical": 1,
    },
    "scrollbarPosition": "content",
    "selectedCellHighlight": false,
    "selectedCellsSpotlight": false,
  },
  "placeholder": "-",
  "showDefaultHeaderActionIcon": false,
  "showSeriesNumber": false,
  "style": Object {
    "cellCfg": Object {
      "height": 30,
      "width": 96,
    },
    "colCfg": Object {
      "height": 30,
      "heightByField": Object {},
      "widthByFieldValue": Object {},
    },
    "collapsedCols": Object {},
    "collapsedRows": Object {},
    "device": "pc",
    "layoutWidthType": "adaptive",
    "rowCfg": Object {
      "heightByField": Object {},
      "width": null,
      "widthByField": Object {},
    },
    "showTreeLeafNodeAlignDot": false,
  },
  "supportCSSTransform": false,
  "tooltip": Object {
    "autoAdjustBoundary": "body",
    "operation": Object {
      "hiddenColumns": false,
      "menus": Array [],
      "sort": false,
      "trend": false,
    },
    "showTooltip": false,
  },
  "totals": Object {},
  "width": 600,
}
`;
